### PR TITLE
Integrating fast_hadamard_transform on C++ level

### PR DESCRIPTION
**Pull request by BlackSamorez**
This PR adds a kernel that wraps together a Fast Hadamard Transform (built from [Tri Dao's implementation](https://github.com/Dao-AILab/fast-hadamard-transform)) and FLUTE GEMM to support [HIGGS](https://arxiv.org/abs/2411.17525). 

Notably, submodule initialization will be needed to build it.

### CODE REVIEW

**Comment by BlackSamorez** [comment]
@HanGuo97 Hi! Works on my machine. Could you check if it compiles and works for you as well?

File: flute/__init__.py [hunk 1]
```diff
@@ -39,6 +39,21 @@
...
+    [
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        int,
+        int,
+        int,
```
Comment (HanGuo97) [inline flute/__init__.py:52]: What does the 3rd integer stands for?

File: flute/csrc/hadamard.cpp
Comment (HanGuo97) [inline flute/csrc/hadamard.cpp:1]: Is this a file a copy of that in the original "fast-hadamard-transform"? (Looks slightly different to me, and if not a copy, what's the difference?)

File: setup.py [hunk 2]
```diff
@@ -89,6 +89,11 @@ def get_extensions() -> List:
     sources = (
         list(glob.glob(os.path.join(extensions_dir, "*.cpp"))) +
         list(glob.glob(os.path.join(extensions_dir, "*.cu"))))
+
```
Comment (HanGuo97) [inline setup.py:92]: minor, could we put them inside `include_dirs` and `sources` definition? (Looks slightly cleaner to me)

File: flute/csrc/qgemm.cpp [hunk 3]
```diff
@@ -369,6 +372,45 @@ qgemm_raw_simple(const at::Tensor& input,
...
+    ).reshape(input_sizes).contiguous();
+}
+
+
+template <
+  typename SMs
+>
+at::Tensor
+qgemm_hadamard(const at::Tensor& input,
+             const at::Tensor& weight,
```
Comment (HanGuo97) [inline flute/csrc/qgemm.cpp:390]: minor indentation inconsistency :)

File: flute/csrc/qgemm.cpp [hunk 4]
```diff
@@ -7,6 +7,9 @@
 #include "cute/numeric/integral_constant.hpp"
 
 
+at::Tensor fast_hadamard_transform(at::Tensor &x, float scale);
```
Comment (HanGuo97) [inline flute/csrc/qgemm.cpp:10]: could we make the signature style similar to others (one line for output type, and one argument per line) --- this is the style used in CUTLASS

File: flute/__init__.py (hunk 1 above)
Comment (HanGuo97) [reply flute/__init__.py:52]: nvm, got it now

**Comment by BlackSamorez** [comment]
Upd: swapped the [Tri Dao's implementation](https://github.com/Dao-AILab/fast-hadamard-transform) for the [HadaCore implementation](https://github.com/pytorch-labs/applied-ai/tree/main/kernels/cuda/inference/hadamard_transform) because the former wouldn't work with `torch.compile` for some batch sizes.
It's ~20% slower though. TODO: choose the better option later.

**Pull request merged**